content/blog/2020-03-06-wasm/index.md

```diff
@@ -0,0 +1,207 @@
+
+<img src="google-earth.png" width="700" >
+
+[earth]: https://www.google.com/earth/
+
+## What is WebAssembly?
+WebAssembly is a binary code format to transfer web applications from the 
+server to the browser. It is incorporated in modern browsers to be used in 
+tandem with existing JavaScript applications, and uses components from 
+existing JavaScript engines to interpret and execute. It has taken the world 
+wide web by storm as
+- it is rolled out by four major browsers (Chrome, Edge, Mozilla and Safari)
+i.e. platform independent
+- it is programming model independent
+- it is hardware independent
+
+What this means is you could write a web application in C, compile it to 
+WebAssembly and use it on any browser on any hardware. This gives much more 
+performance in general than a JavaScript program, which has type 
+classification overheads, performance implications based on which browser 
+you target and how performant is your interpreter.
+
+## Is WebAssembly just C on browsers?
+WebAssembly does brings the performance of pre-compiled languages to web browsers. 
+But it does so side-by-side with traditional Javascript running on the same engines.
+This allows browsers to have best of both worlds, drive the Ferrari on the race track- 
+run performant WebAssembly when it needs performance, or drive the Prius to the 
+grocery store- quickly getting a simple web application up and running.
+
+With WebAssembly, you could use JavaScript for fast development, but also 
+use C where you need performance. You could use C where static typing is
+useful, but JavaScript where dynamic typing is a necessacity for productivity.
+And you could combine these modules, to balance productivity and performance,
+as both these styles can now be executed in the same compiler flow.
+So WebAssembly is more than just C on browsers, it's a carefully coordinated, 
+well-designed, massive engineering effort to reconcile two worlds at odds.
+
+## How does it differ from existing tools?
+So WebAssembly is portable, but doesn't JVM(Java virtual machine) already 
+offer portability? Even though the overall goal has some similarities, Java 
+cannot be run on a browser without plugins and has doesn't have the language
+flexibilty WebAssembly provides. Moreover, validation and memory safety makes 
+WebAssembly more secure from vulnerabilities. Structured control flow and 
+single pass compilation and validation places WebAssembly as a mature option 
+compared to existing bytecode formats such as JVM and CIL (Common Intermediate Language).
+
+WebAssembly also performs better compared to recent technologies with a similar
+goal. Firstly, it can outperform asm.js from optimizations that cannot be done with 
+JavaScript, WebAssembly has faster download with smaller code size, it does not require
+parsing as it's already in a binary format. Moreover, WebAssembly also allows
+certain optimizations to be done statically and to use CPU features not 
+expressible in asm.js. 
```

Comment:
What does it mean to "do an optimization statically"? Do you mean that the wasm *generator* can do optimizations? If so, what is it about wasm that allows this kind of optimization vs. asm.js?

Reply:
I don’t quite follow you here. I’m referring to the tradeoffs between statically typed and dynamically typed, where statically typed is less flexible but offers more opportunities to optimize. I’m not sure what you mean by what allows it to be so?

Reply:
But asm.js is not dynamically typed.

Reply:
Ah, I see what you mean. Thanks.